src/sfizz/Voice.cpp

```diff
@@ -924,12 +929,22 @@ void Voice::Impl::fillWithData(AudioSpan<float> buffer) noexcept
     if (isLooping) {
         int oldIndex {};
         int oldPartitionType {};
+        int blockRestarts { 0 };
         for (unsigned i = 0; i < numSamples; ++i) {
-            int index = (*indices)[i];
-
-            // wrap indices post loop-entry around the loop segment
-            int wrappedIndex = (index <= loop.end) ? index :
-                (loop.start + (index - loop.start) % loop.size);
+            int index = (*indices)[i] - loop.size * blockRestarts;
+            int wrappedIndex = index;
+            // wrap indices post loop-entry around the loop segment and increment the counter if necessary
+            if (index > loop.end) {
+                if (region_->loopCount && loop_.restarts >= *region_->loopCount) {
+                    wrappedIndex = loop.end;
```

Comment:
What about the `sampleEnd` portion after loop end?
this can never get played?

Reply:
Yeah actually I revised it all and also included this part about loop_sustain to handle your comment. Now I find that it behaves differently than sforzando, but we can launch the squad to decide the best behavior.